Add pay payment app integration guide (#730)
centrapay/centrapay-docs: src/content/guides/pay-payment-app-integration.mdx

---
title: Paying a Payment Request
description: Integrate your app with Centrapay to pay payment requests on merchant terminals and POS
nav:
  path: Reference/App Integrations
  order: 1
---

## Overview

This guide is for a [Third-party Asset Provider](/guides/integrating-third-party-asset) to enable paying with their
digital assets via Centrapay.

## Authenticating API calls on behalf of users

Third-party applications can make API calls on behalf of users with Centrapay
API keys. All API requests to Centrapay must be made from the integrator’s backend.
Centrapay API keys must also be managed on the backend and stored as secrets to
minimize the risk of API keys being compromised.

> 

Reviewer: Do we want to add a comment here about rotating API keys?

Author: Do you have a suggestion on what to say? Maybe that's another general guide we should make in the future

Reviewer 2: I think we should have a Managing API Keys guide. Rotating API Keys is important but I don't know that it needs to be mentioned here. Should be part of the certification process though.

Seperate API keys will be used for test and live payments.

> Contact [integrations@centrapay.com](mailto:integrations@centrapay.com) to be issued your API keys.

## Payment Flow Implementation

The payment flow below describes how a Third-party Asset Provider should integrate
with the [Dynamic Merchant QR Code flow](/guides/payment-flows#dynamic-merchant-qr-code) to enable digital payments within their
application. Below is an overview of the process, followed by a more in-depth
look at the integrator’s responsibilities.

> > Note this is assuming the best case scenario, and not accounting for edge cases.<br/><br/>

Reviewer 3: There should be a way to make a multi-line blockquote without adding HTML to our Markdown

Reviewer 3: I'm not sure if this note is providing much value - if we follow this pattern, we'd need to put it on every diagram

Reviewer: I agree I think that it is inferred as part of good design

Author: @tom-blake I think it's valuable to add for now. I think in the future we should have an intro guide that would contain the common principles

Reviewer 3: @IsaacMurtagh Perhaps this is better being expanded on in the testing section instead of introducing noise here? All this callout does is raise more questions.

I am concerned that referring to "edge cases" without documenting them makes our solution look flaky. It's also technical jargon. I'd prefer we remove this line completely and provide useful testing steps for integrators so that they're properly informed

Reviewer 2: I've removed this section. Think the diagram can be inferred to be happy path without extra context.

> You should handle all documented and unknown errors how you see fit and validate if a payment is payable before making requests to avoid a bad user experience.

```mermaid
sequenceDiagram
  autonumber

	actor user
	participant t as Terminal/POS
	participant a as App
	participant b as Integrator's Backend

> 	participant c as Centrapay

Reviewer 3: @IsaacMurtagh These aliases are an anti-pattern that we've tried to avoid in the past - suggest using the full name to make this easier to maintain

> 	user->>t: Select Centrapy

Reviewer 3: ```suggestion
	user->>t: Select Centrapay
```

	t->>c: Create Payment Request
	note over t: Display QR code
	a->>t: Scan QR code
	note over a: Parse Centrapay QR code
	a->>b: Request payment details

> 	b->>c: GET /payment-request/{id}

Reviewer 2: Pay Payment Request has been added

> 	b->>c: GET /payment-request/{id}

Reviewer 2: Also the wording has been updated.

	a->>user: Display Payment Request Details
  user->>a: Approves payment
	a->>b: Initiate payment
	b->>c: POST /payment-request/{id}
	a->>user: Display successful payment
```

> ### User selects Centrapay as their payment method

Reviewer 3: It'd be nice if these headings matched the mermaid autonumbering - is there are reason you're not following the pattern in our other guides?

Reviewer 2: This has been updated to use numbering.

> A payment terminal will provide an option for Centrapay that the user must press.

Reviewer: ```suggestion
The payment terminal will provide an option for Centrapay for the user to select.
```

> A Payment Request will then be created by the merchant and the url field of the

Reviewer: ```suggestion
A Payment Request will then be created for the merchant. Then the url of the
```

> newly created Payment Request will be presented as a QR code.

Reviewer: ```suggestion
Payment Request will be presented as a QR code on the termial.
```

![Concept image of terminal displaying Centrapay QR code](/terminals/displaying-centrapay-qr-code.png)

### User scans the Centrapay QR code via integrated app

> The user will then scan the QR code from inside your application. The valid Centrapay QR Code

Reviewer: ```suggestion
The user will then scan the QR code from inside an integrated application. A valid Centrapay QR Code
```

Reviewer 2: This has been updated to `The user will then scan the QR code from inside the app. Valid Centrapay QR Codes
will have the format`

will have the format `https://app.centrapay.com/pay/{paymentRequestId}`. Validate
the url by asserting it matches the format and ignore any malformed QR codes.

### Get the Payment Request’s details

Once your application has validated the url, use the id of the Payment Request to
call the [Get Payment Request](https://docs.centrapay.com/api/payment-requests#get-a-payment-request)

> API from your backend. The returned [PaymentRequest model](https://docs.centrapay.com/api/payment-requests#payment-request)

Reviewer: ```suggestion
API. The returned [PaymentRequest model](https://docs.centrapay.com/api/payment-requests#payment-request)
```

Reviewer 2: This is now call the 'call the [Get Payment Request](https://docs.centrapay.com/api/payment-requests#get-a-payment-request)
API from your server.' Think it is important to still mention the app should not be calling our API's directly.

> will provide all the details needed to determine if the payment request can be completed.

Reviewer: ```suggestion
will provide the details needed to determine if the payment request can be completed.
```

| Field          | Used for                                                                                                                                                                                                                                                                        |
| -------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| status         | If the status is `new` the Payment Request is able to be paid, if it is `cancelled`, `expired`, or `paid` it is in a final state and payments can no longer be made against the Payment Request                                                                                 |
| paymentOptions | The Payment Options define what Assets a merchant accepts. If your Asset’s type is in the Payment Options array, it is able to be used for payment. If your Asset’s type is not in the Payment Options then the merchant has not been configured to accept your payment method. |

> | liveness       | Determines whether a Payment Request is a test Payment Request. Test Payment Requests should only be payable by test assets, and should not be displayed to real users.                                                                                                         |

Reviewer: ```suggestion
| liveness       | Determines whether a Payment Request is a test Payment Request. Test Payment Requests can only be paid by test assets, and should not be displayed to end users.                                                                                                         |
```

### Pay the Payment Request

> If you’ve determined that the Payment Request is payable with your Asset then you

Reviewer: ```suggestion
If the Payment Request is payable with your Asset then you
```

should display the Payment Request details to the user. Some relevant fields to
display are:

| Field        | Used for                                                                                 |
| ------------ | ---------------------------------------------------------------------------------------- |

> | merchantName | The name of the merchant, user’s should always be made aware of who they are paying.     |

Reviewer: ```suggestion
| merchantName | The name of the merchant, users need to be aware of who they are paying.     |
```

> | value        | The user must be aware of the amount and currency needed to fulfill the Payment Request. |

Reviewer: ```suggestion
| value        | It is esential that the user is aware of the amount and currency needed to fulfill the Payment Request |
```

Once a user has seen the relevant information and confirmed payment you should invoke the
[Pay Payment Request API](https://docs.centrapay.com/api/payment-requests#pay-a-payment-request)
from the backend on behalf of the user. Passing your Third-party asset type and
the authorization property.

If the payment was successful then the status of the payment request will be `paid`
and you can show confirmation of payment to the user.

> ## Payment Extentions

Reviewer 3: ```suggestion
## Payment Extensions
```

then you can allow the user to pay only a portion of the payment request and
then pay the remaining amount with an alternative payment method.

The `partialAllowed` flag on the payment request will determine if the merchant
allows partial payments. If they don’t then you should only allow paying the entire
payment request.

To pay a partial amount of the payment request pass the mode and amount to the pay
payment request API.

| Field  | Used for                                                                                           |
| ------ | -------------------------------------------------------------------------------------------------- |
| mode   | For partial payments the mode should be `partial`                                                    |
| amount | For partial payments you must provide the amount of the Payment Request you want to partially pay. |

The `remainingAmount` property on the payment request can be used to show how much
the user still needs to pay with an alternative method.

## Testing

Centrapay supports test resources that can be used for integration testing. A test

> merchant will  be provided that accepts test payments. If you have supplied a

Reviewer 3: There are a few instances of you/your throughout this guide. We should be consistent throughout our site as to whether we write in second or third person

Author: I think we're writing second person based on the farmlands pos guide.

Reviewer: I don't think that Farmlands is something that we should hold as an example as Farmlands had quite an influence on it. Personally I feel 2nd person feels kind of weird.

Author: A lot of other guides are second person too.

Reviewer 3: Happy with second person - it's more personal. Maybe our designers have an opinion on this. If we go with second person - we need to remove all the third-person language and to directly to the app integrator

Reviewer 2: Leaving the guide as second person for now. I think it reads well and matches the other guides.

test third-party asset then this merchant can be configured to accept it as a
payment method.

The [Centrapay Business Portal](https://app.centrapay.com/business) or APIs can

> be used to mock a real payment terminal to test your application.

Reviewer 3: ```suggestion
be used to test a real payment terminal to test your application.
```

Author: The sentence doesn't make sense with that change but I'll reword

Reviewer 3: Sweet - just wanted to avoid technical terminology so that these guides are accessible to more people
